blog/_posts/2016-12-xx-moredots.md

```diff
@@ -0,0 +1,468 @@
+below.
+
+## Why does Julia need dots to fuse the loops?
+
+You might look at an expression like `2 * X.^2 + 6 * X.^3 - sqrt(X)` and
+think that it is "obvious" that it could be combined into a single loop
+over `X`.  Why can't Julia's compiler be smart enough to recognize this?
+
+The thing that you need to realize is that, in Julia, there is nothing
+particularly special about `+` or `sqrt` — they are arbitrary functions
+and could do *anything*.   `X + Y` could send an email or open
+a plotting window, for all the compiler knows.   To figure out that it
+could fuse e.g. `2*X + Y` into a single loop, allocating a single
+array for the result, the compiler would need to:
+
+1. Once the types of `X` and `Y` are determined, figure out what `*` and `+` functions
+  to call.  (Julia already does this, when type inference succeeds.)
+
+2. Look inside of those functions, realize that they are elementwise loops over `X`
+  and `Y`, and realize that they are [pure](https://en.wikipedia.org/wiki/Pure_function)
+  (e.g. `2*X` has no side-effects like modifying `Y`).
+
+3. Analyze expressions like `X[i]` (which are calls to a function `getindex(X, i)`
+  that is "just another function" to the compiler), to detect that they
+  are memory reads/writes and determine what *data dependencies* they imply
+  (e.g. to figure out that `2*X` allocates a temporary array that can be eliminated).
+
+The second and third steps pose an *enormous challenge*: looking at an arbitrary
+function and "understanding" it at this level turns out to be a very hard
+problem for a computer.  If fusion is viewed as a compiler *optimization*, then the
+compiler is only free to fuse if it can *prove* that fusion *won't change the
+results*, which requires the detection of purity and other data-dependency
+analyses.
+
+In contrast, when the Julia compiler sees an expression like `2 .* X .+ Y`,
+it knows just from the *syntax* (the "spelling") that these are elementwise
+operations, and it *guarantees* that it will *always* fuse into a single
+loop, freeing it from the need to prove purity.  This is what we
+term **syntactic loop fusion**, described in more detail below.
+
+### A halfway solution: Loop fusion for a few operations/types
+
+One approach that may occur to you, and which has been implemented in a
+variety of languages (e.g. [Kennedy & McKinley, 1993](http://dl.acm.org/citation.cfm?id=665526);
+[Veldhuizen, 1995](http://citeseerx.ist.psu.edu/viewdoc/summary?doi=10.1.1.43.248);
+[Lewis et al., 1998](http://citeseerx.ist.psu.edu/viewdoc/summary?doi=10.1.1.46.6627);
+[Chakravarty & Keller, 2001](http://dl.acm.org/citation.cfm?id=507661);
+[Manjikian & Abdelrahman, 2002](http://ieeexplore.ieee.org.libproxy.mit.edu/document/577265/);
+[Sarkar, 2010](http://ieeexplore.ieee.org/document/5389392/);
+[Prasad et al., 2011](http://dl.acm.org/citation.cfm?id=1993517);
+[Wu et al., 2012](http://dl.acm.org/citation.cfm?id=2457490)), is to only
```

Comment:
Yes, I want to move the mention of DSLs like Theano and PyOP2 to here, and ParallelAccelerator is definitely something I should mention.  I didn't realize that Numba does loop fusion, but that seems to have been implemented in Numba in 2015.